AI/VCAI/Pathfinding/AINodeStorage.cpp

```diff
@@ -133,28 +138,120 @@ std::vector<CGPathNode *> AINodeStorage::calculateNeighbours(
 	return neighbours;
 }
 
+void AINodeStorage::setHero(HeroPtr heroPtr)
+{
+	hero = heroPtr.get();
+}
+
+class TownPortalAction : public ISpecialAction
+{
+private:
+	const CGTownInstance * target;
+	const HeroPtr  hero;
+
+public:
+	TownPortalAction(const CGTownInstance * target)
+		:target(target)
+	{
+	}
+
+	virtual Goals::TSubgoal whatToDo(HeroPtr hero) const override
+	{
+		const CGTownInstance * targetTown = target; // const pointer is not allowed in settown
+
+		return sptr(Goals::AdventureSpellCast(hero, SpellID::TOWN_PORTAL).settown(targetTown).settile(targetTown->visitablePos()));
+	}
+};
+
 std::vector<CGPathNode *> AINodeStorage::calculateTeleportations(
 	const PathNodeInfo & source,
 	const PathfinderConfig * pathfinderConfig,
 	const CPathfinderHelper * pathfinderHelper)
 {
 	std::vector<CGPathNode *> neighbours;
-	auto accessibleExits = pathfinderHelper->getTeleportExits(source);
-	auto srcNode = getAINode(source.node);
 
-	for(auto & neighbour : accessibleExits)
+	if(source.isNodeObjectVisitable())
 	{
-		auto node = getOrCreateNode(neighbour, source.node->layer, srcNode->chainMask);
+		auto accessibleExits = pathfinderHelper->getTeleportExits(source);
+		auto srcNode = getAINode(source.node);
 
-		if(!node)
-			continue;
+		for(auto & neighbour : accessibleExits)
+		{
+			auto node = getOrCreateNode(neighbour, source.node->layer, srcNode->chainMask);
+
+			if(!node)
+				continue;
 
-		neighbours.push_back(node.get());
+			neighbours.push_back(node.get());
+		}
+	}
+
+	if(hero->getPosition(false) == source.coord)
+	{
+		calculateTownPortalTeleportations(source, neighbours);
 	}
 
 	return neighbours;
 }
 
+void AINodeStorage::calculateTownPortalTeleportations(
+	const PathNodeInfo & source,
+	std::vector<CGPathNode *> & neighbours)
+{
+	SpellID spellID = SpellID::TOWN_PORTAL;
+	const CSpell * townPortal = spellID.toSpell();
+	auto srcNode = getAINode(source.node);
+
+	if(hero->canCastThisSpell(townPortal) && hero->mana >= hero->getSpellCost(townPortal))
+	{
+		auto towns = cb->getTownsInfo();
```

Comment:
Unfortunately town portal also work include ally towns. Please look at:
```
TownPortalMechanics::getPossibleTowns
```
This really need to be worked around properly.

And just in case please add some comment that it is code duplication of sorts even if Alex still need more refactoring on his code.